lib/SIL/TypeLowering.cpp

```diff
@@ -1565,27 +1565,56 @@ CanType TypeConverter::getLoweredRValueType(AbstractionPattern origType,
   return substType;
 }
 
-const TypeLowering &TypeConverter::getTypeLowering(SILType type) {
+const TypeLowering &TypeConverter::getTypeLowering(SILType type, ResilienceExpansion forExpansion) {
   auto loweredType = type.getASTType();
   auto key = getTypeKey(AbstractionPattern(getCurGenericContext(), loweredType),
                         loweredType);
 
-  return getTypeLoweringForLoweredType(key);
+  return getTypeLoweringForLoweredType(key, forExpansion);
 }
 
 const TypeLowering &
-TypeConverter::getTypeLoweringForLoweredType(TypeKey key) {
+TypeConverter::getTypeLoweringForLoweredType(TypeKey key, ResilienceExpansion forExpansion) {
   auto type = key.SubstType;
   assert(type->isLegalSILType() && "type is not lowered!");
   (void)type;
 
   // Re-using uncurry level 0 is reasonable because our uncurrying
   // transforms are idempotent at this level.  This means we don't
   // need a ton of redundant entries in the map.
-  if (auto existing = find(key))
-    return *existing;
+  const TypeLowering *lowering = find(key);
+  if (!lowering) {
+    lowering = &getTypeLoweringForUncachedLoweredType(key);
+  }
+  assert(lowering->forExpansion == ResilienceExpansion::Minimal &&
+         "the first lowering in the list must be for minimal expansion");
 
-  return getTypeLoweringForUncachedLoweredType(key);
+  if (key.isDependent() || !lowering->isResilient()) {
+    // Don't try to refine the lowering for other resilience expansions if
+    // we don't expect to get a different lowering anyway.
+    return *lowering;
+  }
+
+  // Search for a matching lowering in the linked list of lowerings.
+  while (true) {
```

Comment:
This looks O(n), can we do this better?

Reply:
We could, but currently it's not worth the effort, as n = 1.
In case we add many more resilience expansions in the future we can think of a more complicated lookup mechanism.